File: agent-ts/src/prompts.ts
/**
 * System prompt templates and user context injection.
 *
 * Port of agent/prompts.py. We omit format_history() and _mask_tool_output()
 * since pi-mono handles context management (compaction, branch summaries, etc.).
 */

import { readdirSync, readFileSync } from "node:fs";
import { resolve, extname, basename, join } from "node:path";
import { existsSync, statSync } from "node:fs";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const PERSISTENT_ROOT =
  process.env.ABAX_PERSISTENT_ROOT ?? "/tmp/abax-data";

export const SYSTEM_PROMPT = `\
You are Abax, a helpful assistant with access to a sandboxed execution environment.

You have tools to execute commands, read/write files, list directories, and control \
a browser inside the sandbox. The sandbox runs Python 3.12 with beancount, pandas, \
matplotlib, and Playwright+Chromium pre-installed.

Key paths:
- /workspace/ — working directory for the current task
- /data/ — persistent user data (survives across sessions)

Guidelines:
- Use execute_command for any computation, data processing, or package installation.
- Use file tools for reading/writing data files.
- Use browser tools when you need to interact with web pages.
- Be concise in your responses. Show results, not process.
- If a command fails, diagnose the error and try a different approach.

{user_context}`;

// ---------------------------------------------------------------------------
// User context
// ---------------------------------------------------------------------------

/**
 * Read all .md files from `${PERSISTENT_ROOT}/${userId}/context/`.
 *
 * Returns a formatted string for injection into the system prompt.
 * Performs path traversal validation to stay within PERSISTENT_ROOT.
 */
export function readUserContext(userId: string): string {
  const contextDir = resolve(PERSISTENT_ROOT, userId, "context");

  // Path traversal guard
  if (!contextDir.startsWith(resolve(PERSISTENT_ROOT))) {
    return "";
  }

  if (!existsSync(contextDir) || !statSync(contextDir).isDirectory()) {
    return "";
  }

  const parts: string[] = [];

  const entries = readdirSync(contextDir).sort();
  for (const name of entries) {
    const fullPath = join(contextDir, name);
    if (extname(name) !== ".md") continue;
    try {
      const stat = statSync(fullPath);
      if (!stat.isFile()) continue;
      const content = readFileSync(fullPath, "utf-8");
      const stem = basename(name, ".md");
      parts.push(`## ${stem}\n${content}`);
    } catch {
      // Skip unreadable files
      continue;
    }
  }

  if (parts.length === 0) return "";
  return "User context:\n" + parts.join("\n\n");
}

// ---------------------------------------------------------------------------
// Prompt builder
// ---------------------------------------------------------------------------

/**
 * Build the complete system prompt with user context injected.
 */
export function buildSystemPrompt(userId: string): string {
  const ctx = readUserContext(userId);
  return SYSTEM_PROMPT.replace("{user_context}", ctx);
}

File: agent-legacy/prompts.py
"""System prompt templates and user context injection."""

import os
from pathlib import Path

PERSISTENT_ROOT = os.getenv("ABAX_PERSISTENT_ROOT", "/tmp/abax-data")

SYSTEM_PROMPT = """\
You are Abax, a helpful assistant with access to a sandboxed execution environment.

You have tools to execute commands, read/write files, list directories, and control \
a browser inside the sandbox. The sandbox runs Python 3.12 with beancount, pandas, \
matplotlib, and Playwright+Chromium pre-installed.

Key paths:
- /workspace/ — working directory for the current task
- /data/ — persistent user data (survives across sessions)

Guidelines:
- Use execute_command for any computation, data processing, or package installation.
- Use file tools for reading/writing data files.
- Use browser tools when you need to interact with web pages.
- Be concise in your responses. Show results, not process.
- If a command fails, diagnose the error and try a different approach.

{user_context}"""


def read_user_context(user_id: str) -> str:
    """Read all .md files from {PERSISTENT_ROOT}/{user_id}/context/.

    Returns formatted string for injection into system prompt.
    """
    context_dir = (Path(PERSISTENT_ROOT) / user_id / "context").resolve()
    if not str(context_dir).startswith(str(Path(PERSISTENT_ROOT).resolve())):
        return ""
    if not context_dir.is_dir():
        return ""

    parts = []
    for entry in sorted(context_dir.iterdir()):
        if entry.is_file() and entry.suffix == ".md":
            try:
                content = entry.read_text(encoding="utf-8")
                parts.append(f"## {entry.stem}\n{content}")
            except OSError:
                continue

    if not parts:
        return ""
    return "User context:\n" + "\n\n".join(parts)


def _mask_tool_output(content: str, max_len: int = 200) -> str:
    """Truncate long content from old assistant messages (tool outputs).

    Observation masking (JetBrains Research): keep reasoning intact,
    only truncate verbose tool outputs in older turns.
    """
    if len(content) <= max_len:
        return content
    return content[:max_len] + "\n[output truncated]"


def format_history(
    history: list[dict] | None, keep_recent: int = 6
) -> str:
    """Format conversation history for system prompt injection.

    Observation masking strategy:
    - Recent messages (last `keep_recent`): kept verbatim
    - Older user messages: always kept verbatim (preserve intent)
    - Older assistant messages: long content truncated (tool outputs are verbose)
    """
    if not history:
        return ""
    parts = []
    cutoff = len(history) - keep_recent
    for idx, msg in enumerate(history):
        role = msg.get("role", "user")
        content = msg.get("content", "")
        # Only mask old assistant messages — user messages always kept intact
        if idx < cutoff and role == "assistant":
            content = _mask_tool_output(content)
        parts.append(f"[{role}]: {content}")
    return "Conversation history:\n" + "\n".join(parts) + "\n\nContinue the conversation."


def build_system_prompt(
    user_id: str, history: list[dict] | None = None
) -> str:
    """Build the complete system prompt with user context and history."""
    ctx = read_user_context(user_id)
    hist = format_history(history)
    prompt = SYSTEM_PROMPT.format(user_context=ctx)
    if hist:
        prompt += "\n\n" + hist
    return prompt
